src/lib/storage/index/abstract_table_index.hpp

```diff
@@ -0,0 +1,253 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "table_index_type.hpp"
+#include "types.hpp"
+
+#include "tsl/sparse_map.h"
+
+namespace opossum {
+
+/**
+ * Basic forward iterator type for iteration over RowIDs, e.g. for table indexes. The default implementation of the
+ * virtual properties is meant to represent the iterator of an empty collection. Therefore it shall not be dereferenced
+ * (like and end-iterator), increments do not make a change, and equality comparisons with other instances of this class
+ * always result in true.
+ */
+class BaseTableIndexIterator {
+ public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = void;
+  using difference_type = std::ptrdiff_t;
+  using pointer = void;
+  using reference = RowID;
+
+  BaseTableIndexIterator(const BaseTableIndexIterator& itr) = default;
+  BaseTableIndexIterator() = default;
+  virtual ~BaseTableIndexIterator() = default;
+
+  virtual reference operator*() const { throw std::logic_error("cannot dereference on empty iterator"); }
+  virtual BaseTableIndexIterator& operator++() { return *this; }
+  virtual bool operator==(const BaseTableIndexIterator& other) const { return true; }
+  virtual bool operator!=(const BaseTableIndexIterator& other) const { return false; }
+
+  virtual std::shared_ptr<BaseTableIndexIterator> clone() const { return std::make_shared<BaseTableIndexIterator>(); }
+};
+
+/**
+ * Forward iterator that iterates over a tsl::robin_map which maps a DataType to a vector of RowIDs. The iteration
+ * process is as if the map would have been flattened and then been iterated.
+ *
+ * @tparam DataType The key type of the underlying map.
+ */
+template <typename DataType>
+class TableIndexIterator : public BaseTableIndexIterator {
+ public:
+  using MapIteratorType = typename tsl::sparse_map<DataType, std::vector<RowID>>::const_iterator;
```

Comment:
That's what I meant a few days ago with eliminating the concrete hash map. However, it was not in `PartialHashIndex` but in `AbstractTableIndex`. In this abstract class, we don't want to depend on individual data structures of the concrete index implementation.